### PR TITLE
Autosuggest features

### PR DESCRIPTION
**Pull request by cvalarida**
The connected ticket is a bit ambiguous, so I'll explain more of what's going on here.

This work allows the type ahead field to accept a `getOptions` callback which can be called whenever input is updated. By supplying `setData` to the callback, we can debounce the function so it's not called _every_ keystroke.

We needed this for the facility names because the list is unreasonably long to store everything in the front end. Instead, we'll fetch the new list of options when the user is done typing.

### CODE REVIEW

File: src/js/disability-benefits/526EZ/config/form.js
Comment (cvalarida) [inline src/js/disability-benefits/526EZ/config/form.js]: I didn't want to write two data util functions in one PR, so I stuck with using lodash for merge.

File: src/js/disability-benefits/526EZ/helpers.jsx
Comment (cvalarida) [inline src/js/disability-benefits/526EZ/helpers.jsx]: I'll be writing another ticket to hook this up to the api. I was trying to keep this PR smaller. :man_shrugging:

File: test/common/schemaform/fields/AutosuggestField.unit.spec.jsx
Comment (cvalarida) [inline test/common/schemaform/fields/AutosuggestField.unit.spec.jsx]: Skipping this for now because it's arguably not necessary for MVP, but I want to keep it on our radar.

File: src/js/disability-benefits/526EZ/config/form.js
Comment (jbalboni) [inline src/js/disability-benefits/526EZ/config/form.js]: I kinda feel like this should be in the field component, instead of leaving it to someone creating a form. It's generally a good practice to not fire a request on every key stroke.

Comment (cvalarida) [reply src/js/disability-benefits/526EZ/config/form.js]: Yeah, I left it out here because this callback isn't _necessarily_ an AJAX request. I can't think of any other legitimate use cases, but I opted for the most flexibility.

That said, I'm not opinionated. I can make the debounce wait time a configurable option. If it's then set to 0, it's _basically_ like it's not debounced at all, which solves that problem if anybody really needs it solved.

Good point. I'll move it. :+1:

File: src/applications/common/schemaform/fields/AutosuggestField.jsx
Comment (jbalboni) [inline src/applications/common/schemaform/fields/AutosuggestField.jsx]: What if we returned a string when `freeInput` was true? Is that too disruptive to the rest of the code in this component?

Comment (cvalarida) [reply src/applications/common/schemaform/fields/AutosuggestField.jsx]: :thinking: Good question. I'm not sure. It seems like not a great idea because we'd be returning different data types, but I'd have to try it out to know for sure. (It's been long enough I'm not sure what the cascading effect would be.)

I'll tackle this on Monday.

File: src/applications/disability-benefits/526EZ/helpers.jsx
Comment (cehsu) [inline src/applications/disability-benefits/526EZ/helpers.jsx]: If no label is provided, is a value provided that can be used instead?

File: src/applications/disability-benefits/526EZ/helpers.jsx
Comment (cehsu) [inline src/applications/disability-benefits/526EZ/helpers.jsx]: Do we have any guidance regarding how many of these can be added?

File: src/platform/utilities/data/debounce.js
Comment (cehsu) [inline src/platform/utilities/data/debounce.js]: Can we avoid using the `this` workaround here?

**Review by cehsu — COMMENTED** [review]
Looks good so far. I had a few minor questions/comments.

File: src/applications/common/schemaform/fields/AutosuggestField.jsx
Comment (jbalboni) [reply src/applications/common/schemaform/fields/AutosuggestField.jsx]: I think we already return a string if `enum` is used. I don't mind the data type switching too much, it just has to match the schema. It seems like a win to avoid using this component specific data structure whenever possible. But I don't want to make this more complicated to understand than it already is.

File: src/applications/common/schemaform/fields/AutosuggestField.jsx
Comment (cvalarida) [reply src/applications/common/schemaform/fields/AutosuggestField.jsx]: Yeah. The only time `getOptions` would be defined is if we were querying the api for a list of options. The alternative is specifying an `enum` in the schema.

File: src/applications/common/schemaform/fields/AutosuggestField.jsx
Comment (cvalarida) [reply src/applications/common/schemaform/fields/AutosuggestField.jsx]: Yes. The only time `queryForResults` will be provided / true is if we wanted to query the api for a new list of options every time the field input changes.

File: src/platform/utilities/data/debounce.js
Comment (cvalarida) [reply src/platform/utilities/data/debounce.js]: Oh woops. Good catch. We should really consider typescript or flow at some point...

File: src/applications/disability-benefits/526EZ/helpers.jsx
Comment (cvalarida) [reply src/applications/disability-benefits/526EZ/helpers.jsx]: I'm not sure I understand the question. These are just dummy values returned by our fake api call. I was using these for testing; we'll get rid of them when we hook it up to the api for real.

I'm not sure how many options the autosuggest field can handle, if that's your question. I think I saw an option somewhere to limit the suggestions shown.

File: src/applications/disability-benefits/526EZ/helpers.jsx
Comment (cvalarida) [reply src/applications/disability-benefits/526EZ/helpers.jsx]: I don't think so.

File: src/platform/utilities/data/debounce.js
Comment (cvalarida) [reply src/platform/utilities/data/debounce.js]: I don't think so. I took this pretty much straight from https://medium.com/@griffinmichl/implementing-debounce-in-javascript-eab51a12311e when I realized I was going to have some `this` context issues.

File: src/applications/disability-benefits/526EZ/helpers.jsx
Comment (cehsu) [reply src/applications/disability-benefits/526EZ/helpers.jsx]: Thanks for explaining.

File: src/applications/common/schemaform/fields/AutosuggestField.jsx
Comment (cehsu) [reply src/applications/common/schemaform/fields/AutosuggestField.jsx]: Thanks for explaining.

File: src/applications/common/schemaform/fields/AutosuggestField.jsx
Comment (cehsu) [reply src/applications/common/schemaform/fields/AutosuggestField.jsx]: Thanks for explaining.

File: src/platform/utilities/data/debounce.js
Comment (cehsu) [reply src/platform/utilities/data/debounce.js]: Thanks for the `context`, it's not clear to me what assigning this to context is accomplishing? Could you explain?

File: src/applications/common/schemaform/fields/AutosuggestField.jsx
Comment (cvalarida) [reply src/applications/common/schemaform/fields/AutosuggestField.jsx]: Looks like it shouldn't be a problem to return a string instead. I'll give it a shot.

Thanks!

File: src/applications/common/schemaform/fields/AutosuggestField.jsx
Comment (cvalarida) [reply src/applications/common/schemaform/fields/AutosuggestField.jsx]: Oh! Duh. I remember why I did this. If we don't pass in a `debounceRate` in the `formConfig`, it'll be `undefined`, and that's when I want it to default to 1000. I _also_ want to be able to pass in 0 so it's essentially _not_ debounced.

Comment (cvalarida) [reply src/applications/common/schemaform/fields/AutosuggestField.jsx]: Good news: When I made the change you proposed, it broke tests that it should have broken. Huzzah for testing! :tada:

Comment (cehsu) [reply src/applications/common/schemaform/fields/AutosuggestField.jsx]: That makes sense, thanks for explaining. When would we want a debounce rate of 0? Is it possible to not debounce in that case?

File: src/platform/utilities/data/debounce.js
Comment (cvalarida) [reply src/platform/utilities/data/debounce.js]: Yeah, absolutely! No guarantees it'll be a _good_ explanation, but here goes:

First, we start off by returning a [closure](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Closures) `debouncedFunction`. This will modify the `timeoutId`. Note the use of the `function` keyword here instead of `=>`. What this means is when an object calls the `debouncedFunction`, the `this` context will be the calling object and _not_ `debounce`. That's what we want.

The trouble comes in when we use `setTimeout`. When `setTimeout` calls the function it's passed,  it's a "free function invocation" (I don't know if that's a standard term, but the article I linked to used it), which means the lexical `this` binding is more or less tossed out the window. The function `setTimeout` calls has `this` attached to nothing.

To overcome that, we use [`function.prototype.apply`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function/apply) to give it the `this` we want.

Finally, `setTimeout`'s first parameter takes a function, so we can't just call `setTimeout(func.apply(context, args), wait)`. We need to pass it a function. We use an arrow function here because the `this` binding of the function we pass it doesn't _really_ matter; when it's called, it'll run the function we want with the right context now.

File: src/applications/disability-benefits/526EZ/helpers.jsx
Comment (cehsu) [reply src/applications/disability-benefits/526EZ/helpers.jsx]: Thanks for explaining, does this mean that if there is no label, it is because there is no treatmentCenterName to being with?

File: src/platform/utilities/data/debounce.js
Comment (cehsu) [reply src/platform/utilities/data/debounce.js]: Thanks @cvalarida, using `apply` and a traditional anonymous function (non-arrow) solves the global scope issue with `setTimeout`, I tried logging `this` in the `setTimeout` call with the workaround (storing `this` in the `context `variable): 
```
timeoutId = setTimeout(() => {console.log('this', context); func.apply(context, args);}, wait);
```
and without:
```
timeoutId = setTimeout(() => {console.log('this', this); func.apply(this, args);}, wait);
```
and it printed `AutosuggestField` regardless (in fact, in this case, the method is already bound before we pass it into debounce, so we could pass `null` as `this` in `apply`, and it still works). 

We may also want to .cancel() the `debouncedGetOptions` , so that it does not attempt to change state after the component is unmounted.

File: src/applications/common/schemaform/fields/AutosuggestField.jsx
Comment (cehsu) [reply src/applications/common/schemaform/fields/AutosuggestField.jsx]: It looks like [lodash](https://github.com/lodash/lodash/blob/master/debounce.js) checks for zero wait times and handles those in the next tick when there is no leading call, is that the anticipated use case here?

Comment (cvalarida) [reply src/applications/common/schemaform/fields/AutosuggestField.jsx]: I don't know when we'd want a debounce of 0 except perhaps in tests, but it's added complexity in the form implementation to handle the debouncing there when we're probably always going to want it. Also, making an exception in here to check for 0 and call the non-debounced function seems kind of...misleading, I guess.

I think when `setTimeout` has a `0` wait time, it'll fire as soon as possible, which is what we want, methinks. I'm not sure what problem calling it on the next tick would solve.

File: src/applications/disability-benefits/526EZ/helpers.jsx
Comment (cvalarida) [reply src/applications/disability-benefits/526EZ/helpers.jsx]: Yeah, exactly.

File: src/platform/utilities/data/debounce.js
Comment (cvalarida) [reply src/platform/utilities/data/debounce.js]: Yeah, good point about `.cancel()`. I'll go ahead and add that.

File: src/applications/disability-benefits/526EZ/helpers.jsx
Comment (cehsu) [reply src/applications/disability-benefits/526EZ/helpers.jsx]: Thanks again, does this throw a "cannot read label property of undefined" if no treatmentCenterName is provided?

**Review by cehsu — APPROVED** [review]
LGTM once @jbalboni approves, as he's more familiar with this component.

File: src/applications/common/schemaform/fields/AutosuggestField.jsx
Comment (jbalboni) [inline src/applications/common/schemaform/fields/AutosuggestField.jsx]: It feels like in this case, if we're calling a server for results, all of the matching logic would be offloaded there, and I'm not sure how much of the rest of the method makes sense in that case.

**Review by jbalboni — APPROVED** [review]
Had a question, but looks good overall to me.

File: src/applications/common/schemaform/fields/AutosuggestField.jsx
Comment (cvalarida) [reply src/applications/common/schemaform/fields/AutosuggestField.jsx]: Right, exactly. I want to say previously, it was set up to only work for returning a promise, but I found that wasn't always going to work. It's been a while, so I forget the exact reasons, I'm afraid.

:thinking: Actually, now that I think about it, this might be a vestige of debouncing in the `formConfig` instead of in here. Maybe we don't need it now?

File: src/applications/common/schemaform/fields/AutosuggestField.jsx
Comment (cvalarida) [reply src/applications/common/schemaform/fields/AutosuggestField.jsx]: This wasn't a deliberate decision when I wrote it, but I think the rest of the method is still useful for while we're waiting for `getOptions` to fire (because debounce). While we're still typing, it'll match (or not) using our current list. When we stop typing, it'll fetch the new list based on that input.

**Pull request merged**
